### ISSUE TITLE
Back-end improvements to search results (eg. ranking, highlighting support, facet support)

### ISSUE THREAD

**Issue by jmcmurry** [monarch-initiative/monarch-legacy#1383]
Related to #1382, we need to cover scenarios originally mentioned #1317. We also need behave tests for each of them.

| search issue | query | Top result must be | results must include | status |
| --- | --- | --- | --- | --- |
| **disease groups** | [Charcot-Marie ](https://beta.monarchinitiative.org/search/Charcot-Marie) | Charcot-Marie Tooth Disease (DOID:10595) |  | Fixed |
| **inversions** | [Marie-Charcot](https://beta.monarchinitiative.org/search/Marie-Charcot) | Charcot-Marie Tooth Disease (DOID:10595) |  | Fixed |
| **hyphenations** | [Marie Charcot](https://beta.monarchinitiative.org/search/Marie%20Charcot) | Charcot-Marie |  | Fixed |
| **exact match** | [Abnormality of the eye](https://beta.monarchinitiative.org/search/Abnormality%20of%20the%20eye) | Abnormality of the eye [phenotype] | Abnormality of the eyelid [phenotype] | Fixed |
| **highlighting** | [Abnormality of the eye](https://beta.monarchinitiative.org/search/Abnormality%20of%20the%20eye) |  | Abnormality of the eyelid [phenotype] | Should be **Abnormality of the eye**lid not **Abnormality of the eyelid** |
| **wildcard stemming** | [Pax](https://beta.monarchinitiative.org/search/Pax) |  | Pax6, Pax4, paxillin etc | [see #1313] working except for highlights |
| **punctuation** | [15q13.3](https://beta.monarchinitiative.org/search/15q13.3) | chromosome 15q13.3 microdeletion syndrome |  | Fixed |
| **Species query special processing** | [snca frog](https://beta.monarchinitiative.org/search/snca%20frog) | snca Xenopus (NCBIGene:100038302) | SNCA human (?) see below | Fixed |
| **Rank human genes first** (if lexically equivalent) | [LHX1](https://beta.monarchinitiative.org/search/lhx1) | human lHX1(NCBIGene:3975) |  | Ranked 20th |
| **Include gene name, not just symbol** | [LHX1](https://beta.monarchinitiative.org/search/lhx1) | human lHX1(NCBIGene:3975) |  | Result blob for NCBIGene:3975 should include "LIM homeobox 1" not just "LHX1" |
- [ ] Results must be returned in a way that supports faceting across **all** of the results, not just those on the first page. We can't write a behave test as those counts are likely to change based on the data. Perhaps though we could say category results greater than X? @kltm curious what you think.
  - [ ] Species (lay terms where possible (human, mouse, horse, cattle, chicken, fish, worm, etc)
  - [ ] Search results type: 
    - Disease
    - Disease group
    - Gene
    - Pathway
    - Gene group [species agnostic pending other improvements]
    - Model [pending indexing of genotypes]
    - Literature [pending indexing of literature]; not even sure about this one, feedback welcome
    - Case [eg. UDN pages etc]; again here, not sure this is a useful search facet. Not sure how often people would be searching for a specific case.

Because the search axes in the facets are disjoint by definition, don't worry about supporting checkbox-style or ternary queries unless it is dead easy. We can revisit.

**Comment by jmcmurry** [comment]
cc: @yuanzhou. Please put any comments related to back-end improvements for Jeremy here and I will update the summary table at the top accordingly.

**Comment by cmungall** [comment]
[moved to https://github.com/monarch-initiative/monarch-app/issues/1314#issuecomment-256715926]

btw, awesome grid, v useful to see this laid out.

**Comment by mellybelly** [comment]
agreed +1 on grid. If we can do all of the above that would be awesome! make it so! [moved to #1314 ]

**Comment by jmcmurry** [comment]
> Currently the distinction between disease and disease-group is a pure UI level hack that exists to prevent killer queries

@cmungall @harryhoch do you think it would be better to state the number of subclasses rather than make an arbitrary binary distinction?

**Comment by harryhoch** [comment]
Interesting question. In the ideal sense, Monarch could be very well-suited for helping people ask questions like - "what can differing model systems tell us about the various manifestations of what we call 'Parkinson's disease'". That said, I'm completely with @cmungall - defining these classes is likely to be exceedingly difficult.

@jmcmurry, I assume your question lies in how do we display "disease" vs. "disease groups"? Right now we have different displays that, if anything, fail to make the distinction clear. I like the idea of including the number of subclasses, but I think that we should consider whether we want to have "disease groups" as continuing entities, and, if so, how we might design displays that would make them useful.

**Comment by yuanzhou** [comment]
@jnguyenx Is there a way that I can get all search results without using the number of rows to be returned?

https://github.com/monarch-initiative/monarch-app/blob/master/lib/monarch/api.js#L1053

So basically I'm talking about the `rows` and `start` arguments, I know this is designed with paging in mind, but I'm not quite sure if paging with work out very well with filters since filters apply to the whole result set.

**Comment by pnrobinson** [comment]
?Two small points.

1) Unless you do a molecular diagnosis or at least some additional analysis, you may not know the disease gene, and thus "Charcot-Marie-Tooth disease" is a valid clinical disease diagnosis and not a group (not more a group than 99% of the other diagnoses in medicine that have heterogeneous etiologies).

2) Marie-Charcot is not in use at all -- it is a weird search and we do not need to optimise it imho.

-Peter


Peter Robinson

Professor of Computational Biology

The Jackson Laboratory for Genomic Medicine

10 Discovery Drive

Farmington, CT 06032

860.837.2095 t | 860.990.3130 m

peter.robinson@jax.org<mailto:peter.robinson@jax.org>

www.jax.org



The Jackson Laboratory: Leading the search for tomorrow's cures


________________________________

**Comment by jmcmurry** [comment]
> Marie-Charcot is not in use at all

Totally agreed, this particular use case is somewhat artificial, however there may be other cases where inverting the word order is more natural / legitimate eg skin/hair/eye phenotype. Someone may not remember what the official order is. This is a proxy for that.

> I assume your question lies in how do we display "disease" vs. "disease groups"?

Yes, however, scoped for now to just site search results. I'm always frustrated by not being able to find "fanconi" or "marfan". I'm wondering whether site search results could be grouped with the parent on top? This representation kinda drives me nuts:

<img width="877" alt="screen shot 2016-10-25 at 3 51 23 pm" src="https://cloud.githubusercontent.com/assets/3658863/19707100/7c19e104-9aca-11e6-8c0c-80f89d310c6e.png">

**Comment by cmungall** [comment]
I like the idea of showing hierarchy as part of search - I do this when I query on the command line. But this should probably be an extra enhancement ticket, as there are architectural issues to consider here. We should focus on what can be done with the existing index for this round.

**Comment by jnguyenx** [comment]
@yuanzhou The line of code you pasted is for an owlsim search, you should rather use [that](https://github.com/monarch-initiative/monarch-app/blob/master/lib/monarch/api.js#L2771) for the search page.

Yes, filters will be computed on the whole result set. I think I know what bugs you:
The client will receive only part of the result set (e.g. the fist page to start with). When he'll click on a filter, another solr query will be done and will return a filtered result set. Same for sorting if we'll add it some day. You don't have to manage that on the UI side, sorl will do the work for you. You just need to display what it gives you and let the data drives that.

**Comment by harryhoch** [comment]
a search for Fanconi can lead to some inappropriately long results - see screenshot.
![screenshot](https://cloud.githubusercontent.com/assets/174707/19737226/97e31336-9b80-11e6-9e95-535f2dfcdff8.png)

**Comment by yuanzhou** [comment]
@jnguyenx sorry that I pasted the wrong line number. I did use the one you suggested.

My concern is if we only display part of the result set per page because of using pagination, then applying filters will generate another new set of results instead of filtering the current data set. Because the filters are supposed to be used to narrow down the current data set instead of fetching new data set. In that case, the filters should be search options before we hit that "Go" button. 

My thought is if we use species and category filters, then they should apply to current data set in one page (might be a long page, but that's why we use filters to shorten it) without pagination. This also means the search page should contain all search results once it's loaded. We are just filtering the results on client side.

@jmcmurry @harryhoch any thoughts?

**Comment by jmcmurry** [comment]
To separate concerns, please see [my comment in 1383](https://github.com/monarch-initiative/monarch-app/issues/1386#issuecomment-256427499) about the desired user-facing behavior. The back end must accommodate dynamic count updates over the active set of results, be it filtered, or unfiltered

**Comment by yuanzhou** [comment]
Here is the screenshot of my work to make sure I explained my concern clearly.

<img width="400" alt="screen shot 2016-10-25 at 3 51 23 pm" src="https://cloud.githubusercontent.com/assets/195873/19739080/a3215602-9b87-11e6-8b1c-f184865bad17.PNG">

In this case, the total number of results is 315. And the two filters apply to the whole data set. Imagine we are using pagination, once a user clicks on the filters, we fetch another data set (which the total number of matches must be smaller), this is considered as a new search with some criteria instead of filtering.

**Comment by jmcmurry** [comment]
Ahhh. If you are saying that the total number of results is 315, and should remain 315 regardless of what filters are applied, I completely agree and the back-end must support this.

**Comment by jmcmurry** [comment]
1) The number of total search results (at the top) must change dynamically according to the filters
2) Selection of a facet in the species column must impact the counts in category facets, but not counts in species facets
3) A selection in category facets must impact counts in the species facets but not in category facets

**Comment by julesjacobsen** [comment]
Are we going to cater for people who know what they want to see, but aren't sure where to look? For example I wanted to have the Monarch world-view of an HPO term, so I did a search with the ID:

https://beta.monarchinitiative.org/search/HP:0000452

The correct hit was the third down:

![monarch_search_hp_0000452](https://cloud.githubusercontent.com/assets/1576670/19762357/96c04daa-9c31-11e6-8fe8-60e5dca964df.PNG)

**Comment by jmcmurry** [comment]
@julesjacobsen yes, search by curie is already in the table above; it is better than it was (it used to be zero results) but agreed ranking needs work (reflected in status).

**Comment by pnrobinson** [comment]
Could we use indenting to represent some of the semantics? i.e., Fanconi anemia etc is not indented, but fanca is?

I think we need to schedule a long skype or a F2F to discuss the strategies and what exactly we want the website to show -- count me in for it

**Comment by yuanzhou** [comment]
Do we consider missing taxon info of some search results as a data issue?

**Comment by jmcmurry** [comment]
Yes, it is a data issue; however one that is likely not to get fixed immediately as it is thorny. Moreover, what we build probably does need to account for missing taxon just in case future data snafus anyway.

**Comment by jnguyenx** [comment]
Some updates on the solr analyzers, now exact matches should be better ranked.

Concerning boosting the human taxon, I don't think that's reasonable to do that. The user should use filters or can type `human` or any synonyms in the search box. We should rely on solr raw scoring and not to try fancy things. Tweaking edge cases like this can break other edge cases etc...

For highlighting, this is coming from solr and it only highlight the whole token unfortunately, as far as I know we can't force it to only highlight matchings characters.

**Comment by jmcmurry** [comment]
[hl.fragsize](https://wiki.apache.org/solr/HighlightingParameters) seems to be configurable, no?

**Comment by jnguyenx** [comment]
Yes it's configurable, but that's not what you're looking for in my opinion.

**Comment by jmcmurry** [comment]
As for human/mouse/fish boosting, is your concern that a) boosting will interfere with other ranking methods or b) boosting is unnecessary as filters get you there? If 'a' is your concern, is there a way that the boosting can be applied _only_ within genes of equivalent symbols?

**Comment by jnguyenx** [comment]
It's actually both, in addition to b), you can also type human or any synonyms in the search box to limit the results.

I prefer to stick to solr per field boosting mechanism and rely only on user input, and not try to reorder ranking manually with tricks, that can go very wrong in my experience.

**Comment by kshefchek** [comment]
Is there a place where we could log queries which do not return the best results (semi subjectively)?  For example:
https://beta.monarchinitiative.org/search/ALS
filter on diseases
ALS (Amyotrophic Lateral Sclerosis) does not appear until the 51st result.

**Comment by kltm** [comment]
@kshefchek Neat, and hard to pin down given the way that the tokenizers are working.
I believe that with the more complicated manager and schema layout in the pipeline (e.g. https://github.com/berkeleybop/bbop-manager-golr/issues/4 ), there should be a better result (direct boosted match on ws tokenized field, less boosted matched on edgengram fields).

**Comment by jnguyenx** [comment]
I believe that this particular example is due to ALS being a synonym only.
I hacked around and put some boosts as prove of concepts, until the issue
that Seth mentioned is fixed.

https://github.com/monarch-initiative/monarch-app/blob/master/lib/monarch/api.js#L2788
You can try and play with the boosts locally.

**Comment by kshefchek** [comment]
@jmcmurry @jnguyenx any plans to enable search on our data-graph versus just the ontologies?  Searching for genotypes and variants would be especially useful.

**Comment by jnguyenx** [comment]
I need to talk to @cmungall for this, he prevented me from doing it but we have never finished the discussion.

**Comment by jnguyenx** [comment]
We switched to the data graph earlier this month, it looks like it's working.

**Comment by jmcmurry** [comment]
broke out curie-specific concerns here https://github.com/monarch-initiative/monarch-app/issues/1625
